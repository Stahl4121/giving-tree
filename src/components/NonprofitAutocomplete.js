import React, { useEffect } from 'react';

import firebase from '../firebase';

import Autocomplete from '@material-ui/lab/Autocomplete';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';

export default function NonprofitAutocomplete(props) {
  const [nonprofits, setNonprofits] = React.useState([]);
  const [nonprofitData, setNonprofitData] = React.useState({ name: '', number: '', email: '', url: '', cf_id: props.cfId })

  let db = firebase.firestore();

  let transformedNonprofits = Object.values(nonprofits).map((item) => {
    item.dataLabel = item.name;
    return item;
  })

  // Keep track of whether the 'add nonprofit' panel is open
  const [adding, setAdding] = React.useState(false);

  const addMode = () => { setAdding(true); }

  const cancelAddMode = () => { setAdding(false); }

  const addNonprofit = () => {
    db.collection('nonprofits').doc().set(nonprofitData)
      .then(function () {
        setAdding(false);
      }).catch(function (error) {
        console.error('Error writing draft: ', error);
      })
  }

  // Load the nonprofits
  useEffect(() => {
    let newNonprofits = [];
    db.collection('nonprofits').where('cf_id', '==', props.cfId)
      .get()
      .then(function (querySnapshot) {
        querySnapshot.forEach(function (doc) {
          newNonprofits.push({ id: doc.id, name: doc.data().name })
        })
      }).then(() => setNonprofits(newNonprofits))
      .catch(function (error) {
        console.log("Error getting documents: ", error);
      });
  }, [adding])

  const handleInput = (e) => {
    let newNonprofitData = nonprofitData;
    newNonprofitData[e.target.id] = e.target.value;
    setNonprofitData(newNonprofitData);
  }

  return (
    <div>
      {(nonprofits.length > 0) &&
        <div>
          <Autocomplete
            options={transformedNonprofits}
            getOptionLabel={nonprofits => nonprofits.name}
            autoHighlight
            onChange={props.callback}
            renderInput={params => (
              <TextField
                {...params}
                variant='outlined'
                onChange={props.callback}
                label='Select a nonprofit'
              />
            )}
          />
          <p>
            or
          </p>
        </div>
      }
      <Button
        color='primary'
        variant='contained'
        onClick={addMode}>Add a new nonprofit
      </Button>
      {
        adding &&
        <Container>
          <TextField
            id='name'
            fullWidth
            label='Nonprofit Name'
            onChange={handleInput} />
          <TextField
            id='number'
            fullWidth
            label='Nonprofit Email'
            onChange={handleInput} />
          <TextField
            id='email'
            fullWidth
            label='Nonprofit Phone Number'
            onChange={handleInput} />
          <TextField
            id='url'
            fullWidth
            label='Nonprofit Website'
            onChange={handleInput} />
          <Grid
            container
            direction="row"
            justify="flex-end"
            alignItems="center">
            <Button
              color='primary'
              variant='outlined'
              onClick={cancelAddMode}>
              Cancel
            </Button>
            <Button
              color='primary'
              variant='contained'
              onClick={addNonprofit}>
              Add Nonprofit
            </Button>
          </Grid>
        </Container>
      }
    </div >
  )
}