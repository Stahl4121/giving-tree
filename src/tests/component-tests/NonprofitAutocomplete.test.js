// __tests__/hidden-message.js
// these imports are something you'd normally configure Jest to import for you
// automatically. Learn more in the setup docs: https://testing-library.com/docs/react-testing-library/setup#cleanup
import '@testing-library/jest-dom';
// NOTE: jest-dom adds handy assertions to Jest and is recommended, but not required

import React from 'react';
import { shallow, mount } from 'enzyme';
import { render, fireEvent, screen } from '@testing-library/react';
import NonprofitAutocomplete from '../../components/NonprofitAutocomplete';

const { mockFirebase } = require('firestore-jest-mock');
const routeComponentPropsMock = {
  // add jest.fn() as needed to any of the objects
  history: {},
  location: {},
  match: {}
};

let wrap;
let inst;

beforeEach(() => {
  sessionStorage.clear();
  wrap = shallow(<NonprofitAutocomplete {...routeComponentPropsMock} />)
  inst = wrap.instance();
})

const callback = (event, value) => {
  console.log(value, 'nonprofit_name');
}
const cfId = '4a1MMPzzdmREry9CUV6T'
const error = false;
const helperText = '';
const initialNonprofit = '95CmUGFPnxFUEagA1Mc3'

describe('Nonprofit Autocomplete basics', () => {

  it('Renders correctly', () => {
    const wrap = shallow(<NonprofitAutocomplete
      callback={callback}
      cfId={cfId}
      error={error}
      helperText={helperText}
      initialNonprofit={initialNonprofit} />)
    
    expect(wrap).toMatchSnapshot();
  });
  
})

// test('shows the children when the checkbox is checked', () => {
//   const callback = (event, value) => {
//     console.log(value, 'nonprofit_name');
//   }
//   const cfId = '4a1MMPzzdmREry9CUV6T'
//   const error = false;
//   const helperText = '';
//   const initialNonprofit = '95CmUGFPnxFUEagA1Mc3'
//   render(<NonprofitAutocomplete
//     callback={callback}
//     cfId={cfId}
//     error={error}
//     helperText={helperText}
//     initialNonprofit={initialNonprofit} />)

//   // query* functions will return the element or null if it cannot be found
//   // get* functions will return the element or throw an error if it cannot be found
//   expect(screen.queryByText('PETA')).toBeNull()

//   // the queries can accept a regex to make your selectors more resilient to content tweaks and changes.
//   // fireEvent.click(screen.getByLabelText(/show/i))

//   // .toBeInTheDocument() is an assertion that comes from jest-dom
//   // otherwise you could use .toBeDefined()
//   expect(screen.getByText('PETA')).toBeInTheDocument()
// })