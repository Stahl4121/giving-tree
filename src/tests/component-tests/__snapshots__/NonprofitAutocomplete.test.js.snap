// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nonprofit Autocomplete basics Renders correctly 1`] = `ShallowWrapper {}`;

exports[`Nonprofit Autocomplete shows text fields if no nonprofits Renders correctly 1`] = `ShallowWrapper {}`;

exports[`Nonprofit Autocomplete shows text fields if no nonprofits Shows text fields if there is no nonprofit selected 1`] = `ShallowWrapper {}`;
